packages/repository/test/acceptance/has-many.relation.acceptance.ts

```diff
@@ -99,26 +117,69 @@ describe('HasMany relation', () => {
     customerId: number;
   }
 
-  let customerRepo: EntityCrudRepository<
+  class OrderRepository extends DefaultCrudRepository<
+    Order,
+    typeof Order.prototype.id
+  > {
+    constructor(@inject('datasources.db') protected db: juggler.DataSource) {
+      super(Order, db);
+    }
+  }
+
+  class CustomerRepository extends DefaultCrudRepository<
     Customer,
     typeof Customer.prototype.id
-  >;
-  let orderRepo: EntityCrudRepository<Order, typeof Order.prototype.id>;
-  function givenCrudRepositoriesForCustomerAndOrder() {
-    const db = new juggler.DataSource({connector: 'memory'});
+  > {
+    constructor(
+      @inject('datasources.db') protected db: juggler.DataSource,
+      @hasManyRepository(OrderRepository)
+      public readonly orders: constrainRepositoryFunction<Order>,
```

Comment:
When I was thinking about the design of a repository for a model with a hasMany relation, I thought we would inject the datasource of the related model only and then build the repository factory inside the constructor using a helper function. This does not need a new decorator to implement/maintain/learn by our users. It also makes it trivial to support the case where both models are attached to the same datasource.

```ts
class CustomerRepository extends DefaultCrudRepository<
  Customer,
  typeof Customer.prototype.id
> {
  constructor(
    @inject('datasources.db')
    protected db: juggler.DataSource,
    @inject('datasources.anotherdb', {optional: true})
    protected ordersDataSource: juggler.DataSource = db,
  ) {
    super(Customer, db);

    // A mock-up just to illustrate my point. The factory may need different arguments.
    this.orders = createHasManyRepository<Order>(
      getHasManyMetadata(Customer, "orders"),
      ordersDataSource,
    );
  }
}
```

I think my proposal also removes the need to create a fake datasource - CustomerRepository constructor already has access to all needed information (source model, target model, source repository instance, etc.)

Thoughts?

/cc @raymondfeng

Reply:
@bajtos IMO this approach is sound and I would like us to try it. @raymondfeng @shimks WDYT?

Reply:
I'm fine to have this style as the starting point. The reason to introduce a decorator for `Repository` is to automate the `orders` property population.
```ts
// A mock-up just to illustrate my point. The factory may need different arguments.
    this.orders = createHasManyRepository<Order>(
      getHasManyMetadata(Customer, "orders"),
      ordersDataSource,
    );
```

Reply:
based on @bajtos 's comment, is it possible to inject the whole repository? 
```ts
class CustomerRepository extends DefaultCrudRepository<Customer, typeof Customer.prototype.id> {
  constructor(
    @inject('datasources.db') protected db: juggler.DataSource,
    @inject('repositories.order') protected orderRepo: Repository,
  ) {
  }
}
```